dali/operators/image/convolution/gaussian_blur.cc

```diff
@@ -0,0 +1,305 @@
+    }
+  }
+  return params;
+}
+
+DimDesc ParseAndValidateDim(int ndim, TensorLayout layout) {
+  static constexpr int kMaxDim = 3;
+  if (layout.empty()) {
+    // assuming plain data with no channels
+    DALI_ENFORCE(ndim <= kMaxDim,
+                 make_string("Input data with empty layout cannot have more than ", kMaxDim,
+                             " dimensions, got input with ", ndim, " dimensions."));
+    return {0, ndim, false, false};
+  }
+  // not-empty layout
+  int axes_start = 0;
+  int axes_count = ndim;
+  bool has_channels = ImageLayoutInfo::IsChannelLast(layout);
+  if (has_channels) {
+    axes_count--;
+  }
+  // Skip possible occurrences of 'C' or 'F' at the beggining
+  TensorLayout layout_tmp = layout;
+  while (ImageLayoutInfo::IsChannelFirst(layout_tmp) || VideoLayoutInfo::IsSequence(layout_tmp)) {
+    axes_start++;
+    axes_count--;
+    layout_tmp = layout_tmp.sub(1);
+  }
+  if (!has_channels) {
+    DALI_ENFORCE(!ImageLayoutInfo::HasChannel(layout_tmp),
+                 make_string("Only channel-first or channel-last layouts are supported, got: ",
+                             layout, "."));
+  }
+  DALI_ENFORCE(
+      !VideoLayoutInfo::HasSequence(layout_tmp),
+      make_string("For sequences, layout should begin with 'F' or 'CF', got: ", layout, "."));
+  DALI_ENFORCE(
+      axes_start <= 2,
+      make_string("Found more the one occurrence of 'F' or 'C' axes in layout: ", layout, "."));
+  DALI_ENFORCE(axes_count <= kMaxDim,
+               make_string("Too many dimensions, found: ", axes_count,
+                           " data axes, maximum supported is: ", kMaxDim, "."));
+  return {axes_start, axes_count, has_channels, axes_start != 0};
+}
+
+// axes here is dimension of element processed by kernel - in case of sequence it's 1 less than the
+// actual dim
+template <typename Out, typename In, int axes, bool has_channels>
+class GaussianBlurOpCpu : public OpImplBase<CPUBackend> {
+ public:
+  using Kernel = kernels::SeparableConvolutionCpu<Out, In, float, axes, has_channels>;
+  static constexpr int ndim = Kernel::ndim;
+
+  explicit GaussianBlurOpCpu(const OpSpec& spec, const DimDesc& dim_desc)
+      : spec_(spec), batch_size_(spec.GetArgument<int>("batch_size")), dim_desc_(dim_desc) {}
+
+  bool SetupImpl(std::vector<OutputDesc>& output_desc, const workspace_t<CPUBackend>& ws) override {
+    const auto& input = ws.template InputRef<CPUBackend>(0);
+    int nsamples = input.size();
+    auto nthreads = ws.GetThreadPool().size();
+
+    output_desc.resize(1);
+    output_desc[0].type = TypeInfo::Create<Out>();
+    output_desc[0].shape.resize(nsamples, input.shape().sample_dim());
+
+    params_.resize(nsamples);
+    windows_.resize(nsamples);
+
+    kmgr_.template Resize<Kernel>(nthreads, nsamples);
+
+    for (int i = 0; i < nsamples; i++) {
+      params_[i] = GetSampleParams<axes>(i, spec_, ws);
+      windows_[i].PrepareWindows(params_[i]);
```

Comment:
You're recalculating (likely) the same windows over and over - this can be quite wasteful for large windows.
```suggestion
      auto p = GetSampleParams<axes>(i, spec_, ws);
      if (params_[i] != p) {
        params_[i] = p;
        windows_[i].PrepareWindows(params_[i]);
      }
```

Reply:
It's already done in PrepareWindows.